### PR TITLE
feat: [CDS-56736]: Adding Step for Blue Green Stage Scale Down.

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
In this PR, we are adding a new Blue Green Stage Scale Down Step. This step will scale down the stage environment in BG deployment. The user want to scale down their stage environment in next stage (not in the deployment stage) as they need to perform validation first. Will add this step to plan creator as fill the TODO's as part of other tickets in this epic: [https://harness.atlassian.net/browse/CDS-55822](https://harness.atlassian.net/browse/CDS-55822)

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/46531)

### CODE REVIEW

File: 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStepInfo.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  @Override
+  public StepType getStepType() {
+    return K8sBGStageScaleDownStep.STEP_TYPE;
+  }
+
+  @Override
+  public String getFacilitatorType() {
+    return OrchestrationFacilitatorType.TASK_CHAIN;
```
Comment (abosii-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStepInfo.java:79]: Why is it task chain? Should be OrchestrationFacilitatorType.TASK

**Review by abosii-harness — DISMISSED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStepInfo.java (hunk 1 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStepInfo.java]: Done

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 11 of 17 files at r1, 1 of 2 files at r2, 4 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46531)* status: 19 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by abosii-harness — DISMISSED** [review]

**Review by yogesh-chauhan — DISMISSED** [review]

**Comment by GoelPratyush99** [comment]
trigger messagecheck

**Comment by GoelPratyush99** [comment]
trigger ut3

**Comment by GoelPratyush99** [comment]
trigger ut7

**Comment by GoelPratyush99** [comment]
trigger ut9

**Review by abosii-harness — DISMISSED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/visitor/helpers/cdstepinfo/K8sBGStageScaleDownStepInfoVisitorHelper.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+@OwnedBy(CDP)
+public class K8sBGStageScaleDownStepInfoVisitorHelper implements ConfigValidator {
+  @Override
+  public void validate(Object object, ValidationVisitor visitor) {
+    // Nothing to validate.
+  }
+
+  @Override
+  public Object createDummyVisitableElement(Object originalElement) {
+    return K8sDryRunManifestStepInfo.infoBuilder().build();
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/visitor/helpers/cdstepinfo/K8sBGStageScaleDownStepInfoVisitorHelper.java:26]: should it not return K8sBGStageScaleDownStepInfo ?

**Review by GoelPratyush99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46531)* status: 19 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/visitor/helpers/cdstepinfo/K8sBGStageScaleDownStepInfoVisitorHelper.java` line 26 at r5](https://reviewable.io/reviews/harness/harness-core/46531#gh-1177552061:-NTx42c60sbIziIDrnoY:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/766f820aee4a0711b3b426deac740a39bfef38c8/125-cd-nextgen/src/main/java/io/harness/cdng/visitor/helpers/cdstepinfo/K8sBGStageScaleDownStepInfoVisitorHelper.java#L26)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

should it not return K8sBGStageScaleDownStepInfo ?
</blockquote></details>

Done.

**Review by yogesh-chauhan — APPROVED** [review]

**Review by abosii-harness — APPROVED** [review]

**Comment by GoelPratyush99** [comment]
trigger review

**Pull request merged**